### ISSUE TITLE
sphinx filterparams inerrability 

### ISSUE THREAD

**Issue by kor4ik** [erikkemperman/sphinxcontrib-autodoc-filterparams#1]
hi Erik,
I would like to contribute your brilliant ext of sphinxcontrib_autodoc_filterparams with idea of making it available for multiple functions.
I would like to commit my changes for your review, in case you'll find it reasonable.

**Comment by erikkemperman** [comment]
Can you explain what you want to achieve? I don't really understand your PR to be honest... Please provide an example usage, how you would use this in your sphinx config.

**Comment by kor4ik** [comment]
I did commit the pull request with my changes. Would you be so kind to review it.
In general, It's making initial code inerrable over all functions in the project. 

conf.py - won't be impacted
extensions = [
    'sphinx.ext.autodoc',
    'sphinxcontrib_autodoc_filterparams'
    ]

def sphinxcontrib_autodoc_filterparams(fun='', param=''):
    return not param.isalpha

**Comment by erikkemperman** [comment]
Hi again,

I actually looked at your PR before, but I just don't understand some of your changes... I've added a couple of remarks.

However, I did notice that you added `return_annotation` to some return statements -- but not the main one at the end. I don't really remember why I return `None` there instead of the value passed in. I'll take a closer look when I have some time.

Also, while I was looking at this code now (it's been a while) I noticed that I was using some deprecated methods that were removed in sphinx 4 -- so I'll have to fix that as well.

Finally, you mention "inerrable" but it seems to me that the example you give would work with the current version (although I don't understand why you would give the `fun` and `param` arguments default values -- this function is always invoked with two explicit arguments by the extension.

Please provide or link an example that doesn't work for you with the current version.

Thanks for taking the time to help improve this -- I hope you understand that I feel uncomfortable merging any PR if I don't understand what it does!

**Comment by kor4ik** [comment]
Hi,
Thank you very much for your review comments. I did correct the PR accordingly.

**Comment by erikkemperman** [comment]
I am unable to reproduce the error you mention. Could you please post your conf.py and the python scripts you are passing through autodoc which is causing this problem? Thanks!

**Comment by kor4ik** [comment]
Thank you very much for your support.
Attached below:
[conf_script.zip](https://github.com/erikkemperman/sphinxcontrib-autodoc-filterparams/files/7364522/conf_script.zip)

**Comment by erikkemperman** [comment]
Thanks!

I've imported your files into my project and am able to produce output without errors:

`make text` produces

```
This module contains most of the reusable functions to support test
cases.

class style_rst.UIAction(driver)

   UI Action class contains all ui user action methods.

   static current_time_generator()

      This function is used to generate current time

      Returns:
         it returns current time string in format: "_Y-m-d_H-M-S

   dropdown_select(self, action, locator_properties, locator_type='id', max_time_out=10, text_value=None)

      This method is used to perform dropdown selection according to
      the data source provided action and data

      Usage:
         attribute = <dropdown> | action = <select_value or
         select_index or select_text> | test_data = <user input -
         value/text/index from the dropdown list>

      Parameters:
         * **locator_properties** -- it takes locator string as
           parameter

         * **locator_type** -- it takes locator type as parameter

         * **max_time_out** -- this is the maximum time to wait for
           particular element

         * **text_value** -- user input

         * **action** -- source data action - select_value /
           select_index / select_text

      Returns:
         it returns nothing
```


Ignore the 'style_rst' bits, that's just the package where I put your file locally.

And if I change the callback inside conf.py to `return not param.isalpha()` [*Note the parentheses at the end!!*] 

```
This module contains most of the reusable functions to support test
cases.

class style_rst.UIAction

   UI Action class contains all ui user action methods.

   static current_time_generator()

      This function is used to generate current time

      Returns:
         it returns current time string in format: "_Y-m-d_H-M-S

   dropdown_select(locator_properties, locator_type='id', max_time_out=10, text_value=None)

      This method is used to perform dropdown selection according to
      the data source provided action and data

      Usage:
         attribute = <dropdown> | action = <select_value or
         select_index or select_text> | test_data = <user input -
         value/text/index from the dropdown list>

      Parameters:
         * **locator_properties** -- it takes locator string as
           parameter

         * **locator_type** -- it takes locator type as parameter

         * **max_time_out** -- this is the maximum time to wait for
           particular element

         * **text_value** -- user input

      Returns:
         it returns nothing
```

Now, the `action` parameter is removed from the output, because `not 'action'.isalpha()` returns `False`.

Without the parentheses, by the way, you're just doing `not <method>` which is always False, and the extension will filter out all parameters:

```
This module contains most of the reusable functions to support test
cases.

class style_rst.UIAction

   UI Action class contains all ui user action methods.

   static current_time_generator()

      This function is used to generate current time

      Returns:
         it returns current time string in format: "_Y-m-d_H-M-S

   dropdown_select()

      This method is used to perform dropdown selection according to
      the data source provided action and data

      Usage:
         attribute = <dropdown> | action = <select_value or
         select_index or select_text> | test_data = <user input -
         value/text/index from the dropdown list>

      Returns:
         it returns nothing
```

But in all three cases I don't get the error you mention.  Can you paste all of the output you see when you run Sphinx with "-vv"?

**Comment by erikkemperman** [comment]
PS If other people run into this issue, it would be better to paste the text of your scripts here rather than a zip file -- easier to see what is going on in a single place.

**Comment by kor4ik** [comment]
You are completely right, while using <method> param.isalpha returns explicit statement False
however I was referring to the example in conf.py as below
 def sphinxcontrib_autodoc_filterparams(fun, param):
    return not param.startswith('locator')

which requires iteration over multiple parameters

**Comment by erikkemperman** [comment]
Ah, but that's not what you put in the zip file :-)

I believe you misunderstand the pattern, perhaps. The callback `sphinxcontrib_autodoc_filterparams` is invoked once for each parameter name. So the way you write it here would remove all parameters whose name starts with `'locator'`, but only after several calls to the callback.

Is that the issue, that you would like a single callback for all of the parameters at once?

**Comment by erikkemperman** [comment]
In any event, this would also not trigger the exception that you mentioned in the PR...

**Comment by kor4ik** [comment]
I'm sorry, but still getting below exception 
WARNING: error while formatting signature for core.selenium.lib.ui_actions.UIAction: Handler <function process_signature at 0x000001EE3C67D790> for event 'autodoc-process-signature' threw an exception (exception: 'type' object is not iterable)

Can it be that I'm creating docstrings from scratch every time ?
Meantime, the only fix that worked for me:

def process_signature(app, what, name, obj, options,
                      signature, return_annotation):
    filter_params = **((getattr(app.config, _ext, None)()),)**

**Comment by erikkemperman** [comment]
Hmm, it's weird. The only thing I can think of is that your version of sphinx (specifically the built-in autodoc extension) calls the `process_signature` with a different argument list than what the published version of my extension expects?

Because otherwise, your change makes sure that `filter_params` is a tuple, which is not callable, and you should just fall into this return:

https://github.com/erikkemperman/sphinxcontrib-autodoc-filterparams/blob/39fce022ef3886936b0657438674f1182ef4a0aa/sphinxcontrib_autodoc_filterparams/__init__.py#L42-L44

And before that return, I don't see any statement where the function might try to iterate over a `type`, like your error says.

Can you add a print to the very beginning of this function, and tell me what it says?

```python3
def process_signature(app, what, name, obj, options,
                      signature, return_annotation):
    from sphinx import version_info
    print('Sphinx version', version_info)
    print('Args', app, what, name, obj, options, signature, return_annotation)
    ...
```

**Comment by erikkemperman** [comment]
By the way, I noticed that you have `core.selenium.lib.ui_actions.UIAction` -- are you generating docs for selenium itself? Or is your package depending on selenium and you are trying to filter parameters of an outside package?

**Comment by kor4ik** [comment]
It's my package depending on selenium and I'm trying to filter params out of my package

Sphinx version (4, 2, 0, 'final', 0)

Args <sphinx.application.Sphinx object at 0x000001FD50280910> module core.selenium.lib.ui_actions <module 'core.selenium.lib.ui_actions' from 'C:\\dev\\Vision\\core\\selenium\\lib\\ui_actions.py'> {'members': <sphinx.ext.autodoc._All object at 0x000001FD50C3EA00>, 'undoc-members': True, 'show-inheritance': True} None None
Sphinx version (4, 2, 0, 'final', 0)
Args <sphinx.application.Sphinx object at 0x000001FD50280910> class core.selenium.lib.ui_actions.UIAction <class 'core.selenium.lib.ui_actions.UIAction'> {'members': <sphinx.ext.autodoc._All object at 0x000001FD50C3EA00>, 'undoc-members': True, 'show-inheritance': True} (driver) None

WARNING: error while formatting signature for core.selenium.lib.ui_actions.UIAction: Handler <function process_signature at 0x000001FD50C7D820> for event 'autodoc-process-signature' threw an exception (exception: 'type' object is not iterable)

**Comment by erikkemperman** [comment]
Thanks! So this tells me you are trying to generate selenium docs, probably because of `show-inheritance: True`.

In sphinx 4 they removed some functions that my extension is using. I still think the error message is weird, but in any case you’re going to need sphinx 3.X until I have time to release an update.

```
pip uninstall sphinx
pip install “sphinx<4”
```

**Comment by kor4ik** [comment]
Thank you so much for your time, spent on this case investigation!!!
I will eagerly wait for the update.
